### PR TITLE
Fix mypy type hint for AdaptiveAvgPool2,3d, AdaptiveMaxPool2,3d

### LINKED ISSUE #49918: Mypy error when passing None to nn.AdaptiveAvgPool2d or nn.AdaptiveMaxPool2d
## 🐛 Bug
`nn.AdaptiveAvgPool2d` and `nn.AdaptiveMaxPool2d` support passing `None` to keep the input dim after pooling as shown in the docs like [nn.AdaptiveAvgPool2d](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveAvgPool2d.html) and [nn.AdaptiveMaxPool2d](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveMaxPool2d.html#torch.nn.AdaptiveMaxPool2d). However, `mypy` shows error that it is incompatible type when passing `None`.

## To Reproduce

```python
class Model(nn.Module):
    def __init__(self):
        ...
        self.pool = nn.AdaptiveAvgPool2d((1, None))

    def forward(self, inputs):
        # inputs: [B, C, H, W]
        outputs = self.pool(inputs)
        # outputs: [B, C, 1, W]
        return outputs
```
Mypy Error: ```Argument 1 to "AdaptiveAvgPool2d" has incompatible type "Tuple[int, None]"; expected "Union[int, Tuple[int, ...]]"```

## Expected behavior

No error either `nn.AdaptiveAvgPool2d((1, None))` or `nn.AdaptiveAvgPool2d((None, 1))`

## Environment

 - PyTorch Version: 1.7
 - OS: Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8.5

cc @albanD @mruberry @jbschlosser @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by VinhLoiIT**
Fixes #49918

### CODE REVIEW

File: torch/nn/common_types.py [hunk 1]
```diff
@@ -24,6 +24,15 @@
...
 _size_6_t = _scalar_or_tuple_6_t[int]
 
+# For arguments which represent optional size parameters (eg, adaptive pool parameters)
+_size_any_opt_t = _scalar_or_tuple_any_t[Optional[int]]
+_size_1_opt_t = _scalar_or_tuple_1_t[Optional[int]]
+_size_2_opt_t = _scalar_or_tuple_2_t[Optional[int]]
+_size_3_opt_t = _scalar_or_tuple_3_t[Optional[int]]
+_size_4_opt_t = _scalar_or_tuple_4_t[Optional[int]]
+_size_5_opt_t = _scalar_or_tuple_5_t[Optional[int]]
+_size_6_opt_t = _scalar_or_tuple_6_t[Optional[int]]
```
Comment (rgommers) [inline torch/nn/common_types.py:34]: Can you remove the unused types here (1/4/5/6)?

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @VinhLoiIT. Could you also add your example showing the mypy failure as a regression test towards the end of `test/test_type_hints.py`?

This is going to cross paths with gh-49504, which enables running mypy over `pooling.py` in CI. That PR is about to land, so let's wait for that before merging this one.

File: torch/nn/common_types.py (hunk 1 above)
Comment (VinhLoiIT) [reply torch/nn/common_types.py:34]: Sure.

**Comment by rgommers** [comment]
The new tests failed, something is still off.

**Review by rgommers — APPROVED** [review]
LGTM now. The one CI failure is unrelated. I also tested this in combination with gh-49504, and all tests still pass. So I think this is good to merge.

@walterddr since you're landing gh-49504, could you look at this one as well?

**Comment by walterddr** [comment]
i will take a look. 

FYI type annotation PR will most likely be landed after new year

**Comment by rgommers** [comment]
Thanks @walterddr. And no worries, there's no real hurry.
